### ISSUE TITLE
Use PYTHON_ARGCOMPLETE_OK

### ISSUE THREAD

**Issue by jayvdb** [coala/coala#3887]
coala documentation says `argcomplete` can be used, but it was recently broken: https://github.com/coala/documentation/pull/424

Currently `coalib/parsing/DefaultArgParser.py` has the following

```
    try:  # pragma: no cover
        # Auto completion should be optional, because of somewhat complicated
        # setup.
        import argcomplete
        argcomplete.autocomplete(arg_parser)
    except ImportError:
        pass
```

note that `argcomplete` is not mentioned in `setup.py` or `requirements.txt`, or `test-requirements.txt`

If we are going to keep that chunk of code above, the dependency should be mentioned in at least setup.py, and in test-requirements, and a real test written.

One way to avoid that is to use `argcomplete`'s automatic mode `PYTHON_ARGCOMPLETE_OK`.

**Comment by jayvdb** [comment]
c.f. https://github.com/coala/coala/issues/2914

**Comment by all3fox** [comment]
What I do not like about `PYTHON_ARGCOMPLETE_OK` (i.e. global mode of `argcomplete`) is that it is bash-specific and may even require you to manually supply the autocompletion folder:

For Arch Linux it works with (instead of the expected default `/etc/bash_completion.d`)

```
sudo activate-global-python-argcomplete --dest /usr/share/bash-completion
```

And after that it works *only* for bash (their docs mention this limitation too).

On the other hand, the following simple local-install procedure works for both bash and zsh:

1. Install argcomplete
2. Add `eval "$(register-python-argcomplete coala)"` to either `.bashrc` or `.zshrc` dependng on what you use
3. Restart the shell (or even just resource the configs)

So, for me it is clear that the local method should be used/recommended/documented instead of the global approach. Just like it is [done now][1].

Now, we could add `argcomplete` to `test-requirements.txt` but I fail to see the point. The `try-except` around `argcomplete.autocomplete(arg_parser)` makes it totally optional. And even if we install it as a test-dependency, and even if the user has *bash* and we add `PYTHON_ARGCOMPLETE_OK` to first 1024 bytes of `coala`, the user *still* would have to manually run `activate-global-python-argcomplete` to enable autocompletion. Meaning they still have to read documentation about that.

So I suggest to not use global mode of argcomplete, to leave documentaion as it is, to not add `argcomplete` as a test dependency.

Finally, I though about end-to-end testing of autocompletion: continuous integration would have to have some extra shell scripts that detect type of shell used by continuos integration, manipulate its config files and resource them. That is too much trouble, I suggest not adding any tests too.

Finally, as for unit-tests, argcomplete has a [decent set of those][2] too.

In short, I suggest closing this issue without any changes.

[1]: https://docs.coala.io/en/latest/Users/Shell_Autocompletion.html 
[2]: https://github.com/kislyuk/argcomplete/blob/master/test/test.py

**Comment by jayvdb** [comment]
Very much appreciate your POV on this one, but I disagree with not adding it to `test-requirements.txt`.
I've submitted a patch with reasoning at https://github.com/coala/coala/pull/4164 , and requested that you review it, so we can continue the discussion there.

**Comment by jayvdb** [comment]
Hmm, not 100% sure about this now.

Why not add `PYTHON_ARGCOMPLETE_OK`, which makes it work for those who do want global mode?
Our docs could still recommend the local mode, and only provide instructions for it.

Fedora RPM does all the hard work for me, a bash user...
```
$ rpm -ql python3-argcomplete
/etc/bash_completion.d/python-argcomplete.sh
/usr/bin/activate-global-python-argcomplete
/usr/bin/python-argcomplete-check-easy-install-script
/usr/bin/python-argcomplete-tcsh
/usr/bin/register-python-argcomplete
/usr/lib/python3.6/site-packages/argcomplete
...
```
